src/main/java/net/dv8tion/jda/internal/utils/cache/MemberCacheViewImpl.java

```diff
@@ -96,8 +96,10 @@ public List<Member> getElementsWithRoles(@Nonnull Collection<Role> roles)
         List<Member> members = new ArrayList<>();
         forEach(member ->
         {
-            if (member.getRoles().containsAll(roles))
-                members.add(member);
+            for (Role role : roles)
+                if (!role.isPublicRole()  && !member.getRoles().contains(role))
+                    return;
+            members.add(member);
```

Comment:
The code seems a bit over complicated imo. I think technically it's correct (process next member if one role is neither the pulic role nor does the member have it) but you need to think about it for a second.

A simplier solution imo would be removing the public role from `roles` (in case it is present). Then the rest of the code still does what is supposed to be.

Reply:
I think, like you mentioned, this might work better:
```java
roles = roles.filter(role -> !role.isPublicRole());
forEach(member -> 
{
  if (member.getRoles().containsAll(roles))
    members.add(member);
})
```
That being said, we need to confirm that `containsAll` returns true when provided an empty list.

Reply:
> roles = roles.filter(role -> !role.isPublicRole());

Should be `roles.removeIf` and perhaps be a new list so the original passed list doesn't get modified. Or yeah, List -> Stream -> List. Perhaps the latter is better since it could be a one liner since the performance impact shouldn't be high.

> That being said, we need to confirm that containsAll returns true when provided an empty list.

Tested it right now, it does (as expected).

Reply:
don't mutate the passed in list.
Otherwise, go with this approach.

Reply:
> Edit: For improved performance, you could also check if only the public role (of this guild) is in the list and then return all members directly.

What about this? It might not be needed but it a) shows that the public role is something special and should improve the performance by a lot. From what I saw `elements.valueCollection()` should provide that, shouldn't it?

Reply:
If the only thing in the list is the public role then yeah, you could skip the check.

So I'd remove the public role and see if the length is 0. If it is, return all members. Otherwise, iterate and check